galsim/compound.py

```diff
@@ -645,6 +744,33 @@ def __str__(self):
         s += ')'
         return s
 
+    def _prepareDraw(self):
+        self._orig_obj._prepareDraw()
+        self.SBProfile = galsim._galsim.SBAutoConvolve(self._orig_obj.SBProfile, self._real_space,
+                                                       self._gsparams)
+
+    def shoot(self, n_photons, rng=None):
+        """Shoot photons into a PhotonArray.
+
+        @param n_photons    The number of photons to use for photon shooting.
+        @param rng          If provided, a random number generator to use for photon shooting,
+                            which may be any kind of BaseDeviate object.  If `rng` is None, one
+                            will be automatically created, using the time as a seed.
+                            [default: None]
+        @returns PhotonArray.
+        """
+        # Setup the rng if not provided one.
+        if rng is None:
```

Comment:
That's a nice solution.  Not quite the approach I was thinking of (I was going to make the constructor from another `BaseDeviate` take a pointer rather than `const BaseDeviate&` and have it default to NULL, aka None in python), but this is probably a nicer way to do it.

As for the new bug, it seems that I have been relying on `BaseDeviate(None)` raising an exception, in kind of a subtle way.  I'll have to think a bit to see what the best solution is to not rely on that, since i do think it should behave as you've now implemented it.